### PR TITLE
installs TableGen and uses it to generate diagnostics

### PR DESCRIPTION
**Pull request by cjdb**
TableGen is an effective way to organise and generate things such as diagnostic data and recognised symbols. It's a little bit Clangey at the moment, but that can probably be smoothed out over time, but this is to be expected since the structure and a lot of the code is directly imported from Clang.

### CODE REVIEW

File: include/schreiber/diagnostic_ids.hpp [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+#define LEXSTART
+#include "diagnostics/DiagnosticLexKinds.inc"
+		num_schreiber_lex_diagnostics,
+	};
+#undef DIAG
+
+	// clang-format on
+
+	/// Adds all the Schreiber diagnostics to the diagnostics engine.
+	void add_diagnostics(clang::DiagnosticsEngine& engine) noexcept;
```
Comment (EricWF) [inline include/schreiber/diagnostic_ids.hpp:33]: This is really non-allocating, non-throwing?

File: third_party/llvm/diagnostics/DiagnosticLexKinds.td [hunk 2]
```diff
@@ -0,0 +1,12 @@
+let Component = "Lex", CategoryName = "Lexical issue" in {
+  // Errors (and correstponding notes)
+  def err_unknown_directive : Error<"unknown directive `%0`">;
+
+  // Warnings (and correstponding notes)
+  def warn_undocumented_decl : Warning<
+    "%select{function}0 %1 is not documented"
```
Comment (EricWF) [inline third_party/llvm/diagnostics/DiagnosticLexKinds.td:7]: Why the select?  Is there supposed to be a `|` after the `{function` to allow for an empty option.

File: third_party/llvm/diagnostics/DiagnosticLexKinds.td [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+  >;
+  def note_undocumented_decl : Note<
+    "use `/// \\undocumented` to indicate that %0 should not be documented"
```
Comment (EricWF) [inline third_party/llvm/diagnostics/DiagnosticLexKinds.td:10]: Minor nit, but this assumes the comment form .Would `/**` not work?

File: include/schreiber/diagnostic_ids.hpp [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+#ifndef SCHREIBER_DIAGNOSTIC_IDS_HPP
+#define SCHREIBER_DIAGNOSTIC_IDS_HPP
+
+#include <clang/Basic/Diagnostic.h>
+#include <clang/Basic/DiagnosticIDs.h>
+
+namespace diag {
+	enum class size {
```
Comment (EricWF) [inline include/schreiber/diagnostic_ids.hpp:11]: Pin the underlying type?

File: include/schreiber/diagnostic_ids.hpp [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+	// clang-format off
+	enum id {
+#define DIAG(ENUM, FLAGS, DEFAULT_MAPPING, DESC, GROUP, SFINAE, NOWERROR,      \
+             SHOWINSYSHEADER, SHOWINSYSMACRO, DEFERRABLE, CATEGORY)            \
+  ENUM,
+#define LEXSTART
+#include "diagnostics/DiagnosticLexKinds.inc"
+		num_schreiber_lex_diagnostics,
+	};
+#undef DIAG
```
Comment (EricWF) [inline include/schreiber/diagnostic_ids.hpp:28]: `#undef LEXSTART`?

Comment (cjdb) [reply include/schreiber/diagnostic_ids.hpp:28]: That's handled by `DiagnosticLexKinds.inc`.

File: include/schreiber/diagnostic_ids.hpp (hunk 4 above)
Comment (cjdb) [reply include/schreiber/diagnostic_ids.hpp:11]: Why?

File: third_party/llvm/diagnostics/DiagnosticLexKinds.td (hunk 3 above)
Comment (cjdb) [reply third_party/llvm/diagnostics/DiagnosticLexKinds.td:10]: Good point.

File: include/schreiber/diagnostic_ids.hpp (hunk 1 above)
Comment (cjdb) [reply include/schreiber/diagnostic_ids.hpp]: Done.

File: third_party/llvm/diagnostics/DiagnosticLexKinds.td (hunk 2 above)
Comment (cjdb) [reply third_party/llvm/diagnostics/DiagnosticLexKinds.td]: Removed select for now.

**Pull request merged**